backend/controller/dal/dal.go

```diff
@@ -501,22 +501,51 @@ func (d *DAL) CreateDeployment(ctx context.Context, language string, moduleSchem
 		}
 	}
 
-	// upsert subscriptions
+	// update subscriptions
+	subscriptionNames := []string{}
 	for _, decl := range moduleSchema.Decls {
 		s, ok := decl.(*schema.Subscription)
 		if !ok {
 			continue
 		}
-		err := tx.UpsertSubscription(ctx, sql.UpsertSubscriptionParams{
+		// look for subscribers
+		if _, hasSubscriber := slices.Find(moduleSchema.Decls, func(in schema.Decl) bool {
+			v, ok := in.(*schema.Verb)
+			if !ok {
+				return false
+			}
+			_, subscribes := slices.Find(v.Metadata, func(in schema.Metadata) bool {
+				md, ok := in.(*schema.MetadataSubscriber)
+				if !ok {
+					return false
+				}
+				return md.Name == s.Name
+			})
+			return subscribes
+		}); !hasSubscriber {
+			// Ignore subscriptions without subscribers
+			// This ensures that controllers don't endlessly try to progress subscriptions without subscribers
+			// https://github.com/TBD54566975/ftl/issues/1685
+			//
+			// It does mean that a subscription will reset to the topic's head if all subscribers are removed and then later re-added
+			continue
+		}
+		if err := tx.UpsertSubscription(ctx, sql.UpsertSubscriptionParams{
 			Key:         model.NewSubscriptionKey(moduleSchema.Name, s.Name),
 			Module:      moduleSchema.Name,
 			TopicModule: s.Topic.Module,
 			TopicName:   s.Topic.Name,
 			Name:        s.Name,
-		})
-		if err != nil {
+		}); err != nil {
 			return model.DeploymentKey{}, fmt.Errorf("could not insert subscription: %w", translatePGError(err))
 		}
+		subscriptionNames = append(subscriptionNames, s.Name)
+	}
+	if err := tx.DeleteOldSubscriptions(ctx, moduleSchema.Name, subscriptionNames); err != nil {
+		return model.DeploymentKey{}, fmt.Errorf("could not delete old subscriptions: %w", translatePGError(err))
+	}
+	if err := tx.DeleteSubscribers(ctx, moduleSchema.Name); err != nil {
```

Comment:
Re:
> These db updates are done in the db commit which inserts the deployment. At that time the previous deployment is still live and the new deployment has min_replicas set to 0. If the next commit (ReplaceDeployment) does not complete then the old deployment will be in a bad state (all subscribers stop receiving events).

Is it possible to refactor this code to combine the 2 commits into a single [Postgres Transaction](https://www.postgresql.org/docs/current/tutorial-transactions.html)? That basically turns it into a single atomic operation, so that if something goes wrong with the second step, then the first will not be committed either. I don't know this code very well, so that may well be a bad idea :p

Reply:
I think there's history in this area of code for why we create a deployment in one transaction, and then replace the old deployment with the new deployment in a second commit. I think @wesbillman mentioned when I first joined that this area had recently gone through some hardening to catch glitches?

Reply:
Oh interesting...